### PR TITLE
Validation for provider spec and unit testing.

### LINKED ISSUE #47: Add `Validation` to `ProviderSpec`
# Summary

In order to ensure the correct behaviour of the MCM driver, the `ProviderSpec` validation needs to be implemented. Here the corresponding implementation (https://github.com/onmetal/machine-controller-manager-provider-onmetal/blob/main/api/validation/validation.go) needs to be improved.

## Definition of done

Ensure the correct `ProviderSpec` and ensure the behaviour by providing corresponding unit tests.

### PR DESCRIPTION
**Pull request by Ashughorla**
# Proposed Changes

- Validation of providerSpec
- Unit testing for providerSpec's validation

Fixes #47

### CODE REVIEW

File: pkg/internal/providerspec.go [hunk 1]
```diff
@@ -27,6 +27,7 @@ var (
 	"prefixName": "my-prefix",
 	"rootDisk": {
 		"volumeClassName": "foo",
+		"volumePoolName": "bar",
```
Comment (afritzler) [inline pkg/internal/providerspec.go:30]: We will remove the `volumePoolName`.

File: pkg/api/validation/validation_test.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+			&v1alpha1.ProviderSpec{
+				RootDisk: &v1alpha1.RootDisk{
+					VolumeClassName: "",
+				},
+			},
+			&corev1.Secret{},
+			fldPath,
+			Not(ContainElement(field.Required(fldPath.Child("spec.rootdisk.volumeclassname"), "provider spec's volumeclassname required"))),
+		),
+		Entry("no volumepool name",
```
Comment (afritzler) [inline pkg/api/validation/validation_test.go:109]: Please remove the volumepoolname check.

File: pkg/api/validation/validation_test.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+		Entry("no image",
+			&v1alpha1.ProviderSpec{
+				Image:    "",
+				RootDisk: &v1alpha1.RootDisk{},
+			},
+			&corev1.Secret{},
+			fldPath,
+			Not(ContainElement(field.Required(fldPath.Child("spec.image"), "provider's spec image required"))),
+		),
+		Entry("no rootdisk",
```
Comment (afritzler) [inline pkg/api/validation/validation_test.go:91]: Root disk is actually an optional field. If it is not set we will RAM boot the machine from `image` without an root disk.

File: pkg/api/validation/validation_test.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+			},
+			&corev1.Secret{
+				Data: map[string][]byte{
+					"userData": nil,
+				},
+			},
+			fldPath,
+			Not(ContainElement(field.Required(fldPath.Child("userData"), "corev1 secret userData required"))),
+		),
+		Entry("no kubeconfig in secret",
```
Comment (afritzler) [inline pkg/api/validation/validation_test.go:58]: The `kubeconfig` in the secret will go. Please remove this check.

File: pkg/api/validation/validation_test.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+			},
+			&corev1.Secret{
+				Data: map[string][]byte{
+					"kubeconfig": nil,
+				},
+			},
+			fldPath,
+			Not(ContainElement(field.Required(fldPath.Child("kubeconfig"), "corev1 secret kubeconfig required"))),
+		),
+		Entry("no namespace in secret",
```
Comment (afritzler) [inline pkg/api/validation/validation_test.go:70]: The `namespace` will go from this secret. Please remove this check.

File: pkg/api/validation/validation.go [hunk 6]
```diff
@@ -51,3 +53,37 @@ func validateSecret(secret *corev1.Secret, fldPath *field.Path) field.ErrorList
 
 	return allErrs
 }
+
+func validateOnmetalMachineClassSpec(spec *v1alpha1.ProviderSpec, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	emptyRootDisk := resource.Quantity{}
+	if spec.RootDisk.Size == emptyRootDisk {
```
Comment (afritzler) [inline pkg/api/validation/validation.go:60]: Root disk is an optional field as it can be empty - see other comment.

**Review by afritzler — CHANGES_REQUESTED** [review]

File: pkg/api/validation/validation.go [hunk 7]
```diff
@@ -51,3 +53,37 @@ func validateSecret(secret *corev1.Secret, fldPath *field.Path) field.ErrorList
...
+		allErrs = append(allErrs, field.Required(fldPath.Child("rootdisk"), "root disk is required"))
+	}
+
+	if spec.RootDisk.VolumeClassName == "" {
+		allErrs = append(allErrs, field.Required(fldPath.Child("rootdisk volumeclassname"), "volumeclassname is required"))
+	}
+
+	if spec.RootDisk.VolumePoolName == "" {
```
[the remark this answers is not in the document]
Comment (Ashughorla) [reply pkg/api/validation/validation.go]: Thanks for comment. Addressed.

File: pkg/internal/providerspec.go (hunk 1 above)
Comment (Ashughorla) [reply pkg/internal/providerspec.go]: Addressed.

File: pkg/api/validation/validation_test.go (hunk 2 above)
Comment (Ashughorla) [reply pkg/api/validation/validation_test.go]: Addressed.

File: pkg/api/validation/validation_test.go (hunk 3 above)
Comment (Ashughorla) [reply pkg/api/validation/validation_test.go]: Addressed.

File: pkg/api/validation/validation_test.go (hunk 4 above)
Comment (Ashughorla) [reply pkg/api/validation/validation_test.go]: Addressed.

File: pkg/api/validation/validation_test.go (hunk 5 above)
Comment (Ashughorla) [reply pkg/api/validation/validation_test.go]: Addressed.

File: pkg/api/validation/validation.go (hunk 6 above)
Comment (Ashughorla) [reply pkg/api/validation/validation.go]: Addressed.

File: pkg/api/validation/validation.go [hunk 8]
```diff
@@ -51,3 +52,29 @@ func validateSecret(secret *corev1.Secret, fldPath *field.Path) field.ErrorList
 
 	return allErrs
 }
+
+func validateOnmetalMachineClassSpec(spec *v1alpha1.ProviderSpec, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	if spec.RootDisk.VolumeClassName == "" {
```
Comment (afritzler) [inline pkg/api/validation/validation.go:59]: As `spec.RootDisk` can be `nil` this check might cause a nil dereference.

File: pkg/api/validation/validation.go [hunk 9]
```diff
@@ -51,3 +52,29 @@ func validateSecret(secret *corev1.Secret, fldPath *field.Path) field.ErrorList
...
+
+	if spec.NetworkName == "" {
+		allErrs = append(allErrs, field.Required(fldPath.Child("networkname"), "networkname is required"))
+	}
+
+	if spec.PrefixName == "" {
+		allErrs = append(allErrs, field.Required(fldPath.Child("prefixname"), "prefixname is required"))
+	}
+
+	if len(spec.Labels) == 0 {
```
Comment (afritzler) [inline pkg/api/validation/validation.go:75]: I would make the `labels` filed optional.

**Review by afritzler — CHANGES_REQUESTED** [review]

File: pkg/api/validation/validation.go (hunk 8 above)
Comment (Ashughorla) [reply pkg/api/validation/validation.go]: Validating spec.RootDisk != nil before checking VolumeClassName

**Review by afritzler — APPROVED** [review]

**Pull request merged**